dmriprep/utils/vectors.py

```diff
@@ -1,482 +1,346 @@
+        return np.squeeze(self.gradients[..., -1] > self._b0_thres)
+
+    def normalize(self):
+        if self._normalized:
+            return
+
+        self._bvecs, self._bvals = normalize_gradients(
+            self.bvecs, self.bvals,
+            b0_threshold=self._b0_thres,
+            bvec_norm_epsilon=self._bvec_norm_epsilon,
+            b_scale=self._b_scale)
+        self._normalized = True
+
+    def generate_rasb(self):
+        """Compose RAS+B gradient table."""
+        if self.gradients is None:
+            self.normalize()
+            _ras = bvecs2ras(self.affine, self.bvecs)
+            self.gradients = np.hstack((_ras, self.bvals[..., np.newaxis]))
+
+    def generate_vecval(self):
+        if not self.bvecs:
+            self._bvecs = bvecs2ras(np.linalg.inv(self.affine), self.gradients[..., :-1])
+            self._bvals = self.gradients[..., -1].flatten()
+
+    @property
+    def pole(self):
+        """
+        Check whether the b-vectors cover a full or just half a shell.
+
+        If pole is all-zeros then the b-vectors cover a full sphere.
+
+        """
+        self.generate_rasb()
+        return calculate_pole(self.gradients[..., :-1], bvec_norm_epsilon=self._bvec_norm_epsilon)
+
+    def to_filename(self, path=None, bvecs=None, bvals=None):
+        if path:
+            np.savetxt(str(path), self.gradients,
+                       delimiter='\t', header='\t'.join('RASB'),
+                       fmt=['%.8f'] * 3 + ['%g'])
+        if bvecs:
+            np.savetxt(str(bvecs), self.bvecs.T, fmt='%.6f')
+        if bvals:
+            np.savetxt(str(bvals), self.bvals, fmt='%.6f')
+
+
+def normalize_gradients(bvecs, bvals, b0_threshold=B0_THRESHOLD,
+                        bvec_norm_epsilon=BVEC_NORM_EPSILON, b_scale=True):
     """
-    Create gradient table from bval/bvec.
+    Normalize b-vectors and b-values.
+
+    The resulting b-vectors will be of unit length for the non-zero b-values.
+    The resultinb b-values will be normalized by the square of the
+    corresponding vector amplitude.
 
     Parameters
     ----------
+    bvec : m x n 2d array
+        Raw b-vectors array.
     bvals : 1d array
         Raw b-values float array.
-    bvecs : m x n 2d array
-        Raw b-vectors array.
-    B0_NORM_EPSILON : float
+    b0_threshold : float
         Gradient threshold below which volumes and vectors are considered B0's.
-   
-    Returns
-    -------
-    gtab : Obj
-        DiPy object storing diffusion gradient information.
-    b0_ixs : 1d array
-        Array containing the indices of the b0s.
-    """
-    from dipy.core.gradients import gradient_table
-    from dmriprep.utils.vectors import check_corruption
-
-    # Perform initial corruption check, correcting for obvious format/character encoding anomalies.
-    bvecs, bvals = check_corruption(bvecs, bvals, B0_NORM_EPSILON)
-
-    # Creating the gradient table
-    gtab = gradient_table(bvals, bvecs, b0_threshold=B0_NORM_EPSILON)
```

Comment:
why did you need so much machinery from dipy, if this is just about a couple of numpy arrays?

Reply:
Not sure what you're referring to exactly?